src/opencl-poh-verify/cl_poh_verify.cpp

```diff
@@ -2,16 +2,308 @@
+    char temp_string[35];
+    sprintf(temp_string, "%s", file_name);
+
+    fp = fopen (temp_string, "w");
+    if (fp == NULL) {
+        fprintf(stderr, "Could not create file %s\n", temp_string);
+        exit(-1);
+    }
+
+    for (size_t i = 0; i < num_elems*SHA256_BLOCK_SIZE; ++i) {
+        fprintf(fp, "%hhu ", hashes[i]);
+    }
+    fclose(fp);
+}
+
 
 extern "C" {
 int poh_verify_many(uint8_t* hashes,
                     const uint64_t* num_hashes_arr,
                     size_t num_elems,
                     uint8_t use_non_default_stream)
 {
-    fprintf(stderr, "poh_verify_many not implemented.");
+    pthread_mutex_lock(&clg_ctx_mutex);
+    bool success = false;
+    //save_input(hashes, num_hashes_arr, num_elems);
+
+    if (initialized == false) {
+        success = cl_check_init();
+        DIE(success == false, "OpenCL could not be init");
+        DIE(cl_crypt_init_locked() == false, "cl_crypt_init_locked failed");
+        initialized = true;
+    } else {
+        fprintf(stderr, "cl_poh_verify_many already initialized\n");
+    }
+
+    gpu_ctx_t *cur_ctx = get_gpu_ctx();
+    pthread_mutex_unlock(&clg_ctx_mutex);
+    
+    size_t nr_bytes_hashes = num_elems * SHA256_BLOCK_SIZE * sizeof(uint8_t);
+    size_t nr_bytes_num_hashes_arr = num_elems * sizeof(uint64_t);
+    setup_gpu_ctx(cur_ctx,
+                    hashes,
+                    num_hashes_arr,
+                    num_elems,
+                    nr_bytes_hashes,
+                    nr_bytes_num_hashes_arr);
+
+    if (num_elems == 0) {
+        return 0;
+    }
+    
+    size_t num_threads_per_block = 64;
+    size_t num_blocks = ROUND_UP_DIV(num_elems, num_threads_per_block) * num_threads_per_block;
+    LOG("num_blocks: %zu threads_per_block: %zu nr hashes: %lu\n",
+           num_blocks, num_threads_per_block, num_elems);                     
+                             
+    CL_ERR( clSetKernelArg(poh_verify_kernel, 0, sizeof(cl_mem), (void *)&cur_ctx->in_out_hashes) );
+    CL_ERR( clSetKernelArg(poh_verify_kernel, 1, sizeof(cl_mem), (void *)&cur_ctx->in_num_hashes_arr) );
+    CL_ERR( clSetKernelArg(poh_verify_kernel, 2, sizeof(cl_uint), (void *)&cur_ctx->in_num_elems) );
+
+	perftime_t start, end;
+    get_time(&start);
+
+    size_t globalSize[2] = {num_blocks * num_threads_per_block, 0};
+    size_t localSize[2] = {num_threads_per_block, 0};    
+    cl_int ret = clEnqueueNDRangeKernel(cmd_queue, poh_verify_kernel, 1, NULL,
+        globalSize, localSize, 0, NULL, NULL);
+    CL_ERR( ret );
+
+    ret = clFinish(cmd_queue);
+    CL_ERR( ret );
+
+    ret = clEnqueueReadBuffer(cmd_queue, cur_ctx->in_out_hashes, CL_TRUE, 0, nr_bytes_hashes, hashes, 0, NULL, NULL);
+    CL_ERR( ret );
+
+    get_time(&end);
+    LOG("time diff: %f\n", get_diff(&start, &end));
+    //save_out(hashes, num_elems);
+    release_gpu_ctx(cur_ctx);
+
+    return 0;
+}
+
+static int
+get_checked_scalar(unsigned char* scalar, const unsigned char* signature) {
```

Comment:
These should go in the ed25519 part of the CL code

Reply:
Done.